src/com/ms/silverking/cloud/dht/management/aws/Util.java

```diff
@@ -21,377 +21,373 @@
-        return isReachableAndStatusIsOk(statusSummary);
-    }
-    
-    private static boolean isReachableAndStatusIsOk(InstanceStatusSummary statusSummary) {
-        return reachabilityPassed(statusSummary.getDetails()) && statusIsOk(statusSummary.getStatus());
-    }
-    
-    private static boolean reachabilityPassed(List<InstanceStatusDetails> details) {
-        for (InstanceStatusDetails detail : details)
-            if (detail.getName().equals("reachability") && detail.getStatus().equals("passed"))
-                return true;
+  static boolean debugPrint = false;
 
-        return false;
-    }
-    
-    private static boolean statusIsOk(String status) {
-        return status.equals("ok");
-    }
-    
-    private static Instance getInstance(String id, List<Instance> instances) {
-        for (Instance instance : instances)
-            if (instance.getInstanceId().equals(id))
-                return instance;
-        
-        throw new RuntimeException("instance '"+id+"' not found");
+  private static final String KEY_PAIR_PREFIX = "sk_key";
+
+  static void printInstance(Instance instance) {
+    if (debugPrint)
+      System.out.printf(
+          "Found instance with id %s, " + "AMI %s, " + "type %s, " + "state %s " + "and monitoring state %s%n",
+          instance.getInstanceId(), instance.getImageId(), instance.getInstanceType(), instance.getState().getName(),
+          instance.getMonitoring().getState());
+  }
+
+  static void printNoDot(String text) {
+    printHelper(text, "");
+    System.out.println();
+  }
+
+  public static void print(String text) {
+    printHelper(text, "......");
+  }
+
+  static void printHelper(String text, String spacer) {
+    System.out.printf("%-65s %s ", text, spacer);
+  }
+
+  public static void printDone(List<String> values) {
+    printDone(String.join(", ", values));
+  }
+
+  public static void printDone() {
+    printDone("");
+  }
+
+  public static void printDone(String value) {
+    System.out.println("done (" + value + ")");
+  }
+
+  static void debugPrint(String text) {
+    if (debugPrint)
+      System.out.println(text);
+  }
+
+  static List<String> getInstanceIds(List<Instance> instances) {
+    List<String> ids = new ArrayList<>();
+
+    for (Instance instance : instances)
+      ids.add(instance.getInstanceId());
+
+    return ids;
+  }
+
+  static List<String> getIps(List<Instance> instances) {
+    List<String> ips = new ArrayList<>();
+
+    for (Instance instance : instances)
+      ips.add(instance.getPrivateIpAddress());
+
+    return ips;
+  }
+
+  static void writeToFile(String filename, String content) {
+    File file = new File(filename);
+
+    try {
+      // mkdirs and createNewFile only create iff those paths/file don't already exist. so no need to check for
+      // existence beforehand.
+      file.getParentFile().mkdirs();
+      file.createNewFile();
+      FileWriter writer = new FileWriter(file);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *writer*. The resource is referenced in statements at the following line: 108. The resource closure statement is at line: 109. There are other execution paths that do not contain closure statements, for example, when Writer.write throws an exception. Either a) close *writer* in a try-finally block or b) close the resource by declaring *writer* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).